Allow year in OWASP string for scalar type

ben-elttam opened the pull request:
- The regex for scalar strings and lists are different.
- Before this change scalar strings did not allow specifying a year, whilst lists must have a year.
- After this change a single string may (optionally) have a year, list behaviour is unchanged, a year must be specified.
- I also tweaked the string regex to allow multiple whitespace characters like the list regex.

I didn't changing the lists regex to make years optional (e.g. use the same regex for strings and lists of strings). I figured when specifying multiple OWASP references, it's probably the one category mapped to different years, rather than several categories from a single year. I think years been mandatory is probably a good thing.

ben-elttam committed:
Allow year in OWASP string for scalar type

- The regex for scalar strings and lists are different.
- Before this change scalar strings did not allow specifying a year,
  whilst lists must have a year.
- After this change a single string may (optionally) have a year, list
  behaviour is unchanged, a year must be specified.

0xDC0DE reviewed (approved):
Thanks for your contribution!

ben-elttam committed:
Update metadata-owasp regex

- I noticed original rule's regex cared about leading zeros.
- Restored old behaviour of leading zeros.
- Scalar string and list now have consistent behaviour of caring about
  leading zeros when year is specified.
- Scalar string leading year is optional when no year is specified.
- Updated test cases.

ben-elttam committed:
metadata-owasp more tests

- Old regex allowed double-zero with year

ben-elttam commented:
I couldn't reproduce the returntocorp/semgrep:1.0.0 failure, it works fine for me.

Looking into the rules I noticed that in the list regex cared about leading zeros. So, I've updated the tests cases and regexs to have matching consistent behaviour of caring about leading zeros. I also noticed the old regex allowed double zero as in `A00`, so I've tightened that up.

I've added more test cases for these.

0xDC0DE commented:
Don't worry about the historical tests. They are not required, and I believe not cause by your PR at the moment.
